File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Persistence/CdmFolder/DataPartitionPatternPersistence.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Persistence.CdmFolder
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Persistence.CdmFolder.Types;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Newtonsoft.Json.Linq;
    using System;
    using System.Collections.Generic;
    using System.Text.RegularExpressions;

    class DataPartitionPatternPersistence
    {
        public static CdmDataPartitionPatternDefinition FromData(CdmCorpusContext ctx, JToken obj)
        {
            var newPattern = ctx.Corpus.MakeObject<CdmDataPartitionPatternDefinition>(CdmObjectType.DataPartitionPatternDef, (string)obj["name"]);
            newPattern.RootLocation = (string)obj["rootLocation"];

            if (obj["globPattern"] != null)
            {
                newPattern.GlobPattern = (string)obj["globPattern"];
            }

            if (obj["regularExpression"] != null)
            {
                newPattern.RegularExpression = (string)obj["regularExpression"];
            }

            if (obj["parameters"] != null)
            {
                newPattern.Parameters = obj["parameters"].ToObject<List<string>>();
            }

            if (obj["lastFileStatusCheckTime"] != null)
            {
                newPattern.LastFileStatusCheckTime = DateTimeOffset.Parse(obj["lastFileStatusCheckTime"].ToString());
            }

            if (obj["lastFileModifiedTime"] != null)
            {
                newPattern.LastFileModifiedTime = DateTimeOffset.Parse(obj["lastFileModifiedTime"].ToString());
            }

            if (obj["explanation"] != null)
            {
                newPattern.Explanation = (string)obj["explanation"];
            }

            if (obj["specializedSchema"] != null)
            {
                newPattern.SpecializedSchema = (string)obj["specializedSchema"];
            }

            Utils.AddListToCdmCollection(newPattern.ExhibitsTraits, Utils.CreateTraitReferenceList(ctx, obj["exhibitsTraits"]));

            return newPattern;
        }

        public static DataPartitionPattern ToData(CdmDataPartitionPatternDefinition instance, ResolveOptions resOpt, CopyOptions options)
        {
            var result = new DataPartitionPattern
            {
                Name = instance.Name,
                LastFileStatusCheckTime = TimeUtils.GetFormattedDateString(instance.LastFileStatusCheckTime),
                LastFileModifiedTime = TimeUtils.GetFormattedDateString(instance.LastFileModifiedTime),
                Explanation = instance.Explanation,
                RootLocation = instance.RootLocation,
                GlobPattern = instance.GlobPattern,
                RegularExpression = instance.RegularExpression != null ? instance.RegularExpression.ToString() : null,
                Parameters = instance.Parameters,
                SpecializedSchema = instance.SpecializedSchema,
                ExhibitsTraits = CopyDataUtils.ListCopyData(resOpt, instance.ExhibitsTraits, options)
            };

            return result;
        }
    }
}

File: objectModel/Python/cdm/persistence/cdmfolder/data_partition_pattern_persistence.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

import dateutil.parser

from cdm.enums import CdmObjectType
from cdm.objectmodel import CdmCorpusContext, CdmDataPartitionPatternDefinition
from cdm.utilities import CopyOptions, ResolveOptions, time_utils, copy_data_utils

from . import utils
from .types import DataPartitionPattern


class DataPartitionPatternPersistence:
    @staticmethod
    def from_data(ctx: CdmCorpusContext, data: DataPartitionPattern) -> CdmDataPartitionPatternDefinition:
        pattern = ctx.corpus.make_object(CdmObjectType.DATA_PARTITION_PATTERN_DEF, data.name if data.get('name') else None)

        pattern.root_location = data.rootLocation
        if data.globPattern:
            pattern.glob_pattern = data.globPattern
        if data.regularExpression:
            pattern.regular_expression = data.regularExpression
        pattern.parameters = data.get('parameters')
        pattern.explanation = data.explanation
        pattern.specialized_schema = data.specializedSchema

        if data.get('lastFileStatusCheckTime'):
            pattern.lastFileStatusCheckTime = dateutil.parser.parse(data.lastFileStatusCheckTime)

        if data.get('lastFileModifiedTime'):
            pattern.lastFileModifiedTime = dateutil.parser.parse(data.lastFileModifiedTime)

        if data.get('exhibitsTraits'):
            exhibits_traits = utils.create_trait_reference_array(ctx, data.exhibitsTraits)
            pattern.exhibits_traits.extend(exhibits_traits)

        return pattern

    @staticmethod
    def to_data(instance: CdmDataPartitionPatternDefinition, res_opt: ResolveOptions, options: CopyOptions) -> DataPartitionPattern:
        data = DataPartitionPattern()

        data.name = instance.name
        data.lastFileStatusCheckTime = time_utils._get_formatted_date_string(instance.last_file_status_check_time)
        data.lastFileModifiedTime = time_utils._get_formatted_date_string(instance.last_file_modified_time)
        data.explanation = instance.explanation
        data.rootLocation = instance.root_location
        data.globPattern = instance.glob_pattern
        data.regularExpression = instance.regular_expression
        data.parameters = instance.parameters
        data.specializedSchema = instance.specialized_schema
        data.exhibitsTraits = copy_data_utils._array_copy_data(res_opt, instance.exhibits_traits, options)

        return data
